### DISCUSSION TITLE
Loss stuck while training a Tacotron2 model on custom dataset in Hindi

### DISCUSSION THREAD

**Discussion by omkarade** [coqui-ai/TTS#1837] [General]
I am training a Tacotron2 model with my custom dataset in Hindi.
Dataset Details: 14hours of data, 22 KHz, 16 bit, Single male Speaker[5sec -13sec]

Issue: the loss seems to be stuck. EvalAudios/audio also noise but word can understand  , but audios generated from test sentences are very noise and not understandable.
Question: Should I train more epochs or is there something else or I change model
my data is noise free
![Screenshot (351)](https://user-images.githubusercontent.com/90675376/183296477-ba09c290-d59e-493c-8013-68b9091e0d5b.png)
![Screenshot (352)](https://user-images.githubusercontent.com/90675376/183296480-55255c57-f169-45a4-9d7c-96145dce231e.png)
![Screenshot (361)](https://user-images.githubusercontent.com/90675376/183296515-9f35f01e-ad5a-4b24-b618-ef66ca281058.png)
![Screenshot (362)](https://user-images.githubusercontent.com/90675376/183296540-22f9679d-7b09-4f78-81ab-98128dd2d298.png)
![Screenshot (358)](https://user-images.githubusercontent.com/90675376/183296562-d02ef20e-3c65-4e0a-bb6a-adde3a1b13b6.png)
![Screenshot (353)](https://user-images.githubusercontent.com/90675376/183297419-a7f918ea-50de-4c79-9315-3e135d68fcc3.png)
![Screenshot (354)](https://user-images.githubusercontent.com/90675376/183297448-5d03ee33-b01d-4c59-9486-022b63b4d9ab.png)

**Comment by omkarade** [comment]
My configuration -----
 
config = Tacotron2Config( 
 output_path=output_path,
 logger_uri=None,
 run_name='run',
 project_name='A',
 run_description='a', 
 print_step=234, 
 plot_step=100, 
 audio=audio_config,
 model_param_stats=False, 
 wandb_entity=None, 
 dashboard_logger='tensorboard', 
 log_model_step=None, 
 save_step=8000, 
 print_eval=False, 
 test_delay_epochs=-1, 
 run_eval=True, 
 distributed_backend='nccl',
 distributed_url='tcp://localhost:54321', 
 mixed_precision=False, 
 epochs=1000,
 batch_size=32, 
 eval_batch_size=8, 
 grad_clip=5.0,
 scheduler_after_epoch=False, 
 lr=0.0001,
 optimizer='RAdam', 
 optimizer_params= {"betas": [0.9,0.998],"weight_decay": 1e-06}, 
 lr_scheduler='NoamLR', 
 lr_scheduler_params={"warmup_steps":4000},
 num_loader_workers=4, 
 num_eval_loader_workers=4,
 use_noise_augment=False, 
 use_phonemes=False, 
 #use_espeak_phonemes= True,
 phonemizer=None,
 #phoneme_language=hi, 
 compute_input_seq_cache=False, 
 text_cleaner='hindi_cleaners', 
 enable_eos_bos_chars=False,
 test_sentences_file='',
 characters=character_config,
 add_blank=False, 
 batch_group_size=0, 
 loss_masking=True,
 #sort_by_audio_len=False,
 #min_audio_len=1,
 #max_audio_len=1000000, 
 #min_text_len=1, 
 #max_text_len=inf, 
 compute_f0=False,
 compute_linear_spec=False,
 use_language_weighted_sampler=False,
 use_gst=False,
 gst=None, 
 gst_style_input=None,
 num_speakers=1, 
 num_chars=127,
 r=2,
 gradual_training=None,
 memory_size=-1,
 prenet_type='original',
 prenet_dropout=True,
 prenet_dropout_at_inference=False,
 stopnet=True,
 separate_stopnet=True,
 stopnet_pos_weight=10.0,
 max_decoder_steps=500, 
 encoder_in_features=512,
 decoder_in_features=512,
 decoder_output_dim=80,
 out_channels=80,
 attention_type='dynamic_convolution',
 attention_heads=4,
 attention_norm='sigmoid',
 attention_win=False,
 windowing=False, 
 use_forward_attn=False,
 forward_attn_mask=False,
 transition_agent=False,
 location_attn=True, 
 bidirectional_decoder=False,
 ddc_r=6,
 use_speaker_embedding=False,
 speaker_embedding_dim=512,
 use_d_vector_file=False,
 d_vector_file=False,
 d_vector_dim=None,
 seq_len_norm=False,
 decoder_loss_alpha=0.25,
 postnet_loss_alpha=0.25,
 postnet_diff_spec_alpha=0.25,
 decoder_diff_spec_alpha=0.25,
 decoder_ssim_alpha=0.25,
 postnet_ssim_alpha=0.25,
 ga_alpha=5.0,

**Reply by thorstenMueller** [reply]
Eval alignment is looking good, but alignment for test phrases not. Are test sentences compareable with test phrases in length and structure? 48k steps is not much training. Maybe more training helps.

**Reply by omkarade** [reply]
loss is still constant.
  --> STEP: 27/234 -- GLOBAL_STEP: **104130**


  --> EVAL PERFORMANCE
     | > avg_loader_time: 0.00259 (-0.00065)
     | > avg_decoder_loss: 16.03601 (-0.10252)
     | > avg_postnet_loss: 17.01884 (-0.09623)
     | > avg_stopnet_loss: 0.29321 (-0.00784)
     | > avg_ga_loss: 0.00054 (-0.00015)
     | > avg_decoder_diff_spec_loss: 0.35906 (+0.00026)
     | > avg_postnet_diff_spec_loss: 0.29674 (-0.00084)
     | > avg_decoder_ssim_loss: 0.67794 (-0.00284)
     | > avg_postnet_ssim_loss: 0.66857 (-0.00257)
     | > avg_loss: 9.06018 (-0.05980)
     | > avg_align_error: 0.50825 (-0.02271)

**Reply by WeberJulian** [reply]
It's not what he asked, he asked about test sentences. It looks like test_sentences or test_sentences_file were not specified so it will default to english test sentence which will break the inference.

**Reply by omkarade** [reply]
i given test sentence are present in train data.

**Comment by WeberJulian** [comment]
Can you post a few samples of your data ?

**Reply by omkarade** [reply]
https://soundcloud.com/omkr3345/sets/sample-1/s-OpYUAnxkSl6?si=1ccb0f19d9154171afcbc294063d4b30&utm_source=clipboard&utm_medium=text&utm_campaign=social_sharing

**Reply by omkarade** [reply]
https://soundcloud.com/omkr3345/sets/sample-3?si=8e547d246f7349cfb4c618f75210a807&utm_source=clipboard&utm_medium=text&utm_campaign=social_sharing

**Reply by WeberJulian** [reply]
The G22 sample had reverb which can hurt the model a lot.
How good are the transcripts?

**Reply by omkarade** [reply]
I merge 2-3 sec wav files. but when merging wav files I add small silence 
code is--
one_sec_segment = AudioSegment.silent(duration=120)

**Reply by omkarade** [reply]
> How good are the transcripts?

I use import speech_recognition as sr

**Reply by omkarade** [reply]
> 120

is silence also effect on model??

**Reply by WeberJulian** [reply]
Oh you used speech recognition for the transcript ? It's pretty bad then. Check if the quality of the transcript is good.

**Reply by WeberJulian** [reply]
trim the trailing silences, it can be really bad for tacotron

**Reply by omkarade** [reply]
or should I  train glowtts ??

**Comment by omkarade** [comment]
now GLOBAL_STEP are  126360 but still loss's is constant 
or should I train Hifi-GAN using Tacotron2 model.is it improve my model??

**Reply by WeberJulian** [reply]
The vocoder is only gonna improve the audio quality, if produced speech is intelligible, it's not gonna help.

**Reply by omkarade** [reply]
or should I train glowtts ??

**Reply by WeberJulian** [reply]
Yeah, it's more robust. You can start there to evaluate your dataset

**Reply by omkarade** [reply]
glowtts  or vits which one is best for my data?

**Reply by WeberJulian** [reply]
VITS is better in terms of audio quality but costs more to train

**Reply by omkarade** [reply]
thanks👍

**Reply by omkarade** [reply]
@WeberJulian punctuations is also matter??

**Reply by WeberJulian** [reply]
yeah

**Reply by omkarade** [reply]
it is not possible to add punctuation in my transcript. Punctuation really matter? or should I train model without Punctuation?

**Reply by WeberJulian** [reply]
If you want your TTS model to sound natural with punctuation, then you need them.
You can aslo find models that punctuate text for you on hugging face

**Reply by omkarade** [reply]
@WeberJulian @thorstenMueller  my text data is clean it is ok to use this cleaner -->
def hindi_cleaners(text):
      return text

**Reply by manmaynakhashi** [reply]
@omkarade you can use that , use espeak phoneme for better result

**Comment by manmaynakhashi** [comment]
```text_cleaner="hindi_cleaners",
    use_phonemes=True,
    phonemizer="espeak",
    phoneme_language="hi",
    phoneme_cache_path=os.path.join(output_path, "commentry_cache"),
```
add this to you config and 
install espeak using sudo apt-get install espeak

**Reply by omkarade** [reply]
how install espeak on google colab notebook ??

**Reply by omkarade** [reply]
@manmaynakhashi  hello sir ! 
you train any Hindi model???

**Reply by sudip550** [reply]
> @manmaynakhashi hello sir ! you train any Hindi model???

Did you done sir

**Comment by omkarade** [comment]
@WeberJulian @thorstenMueller  I need increase max_decoder_steps ?it's on 500 ?
 > Decoder stopped with `max_decoder_steps` 500

**Reply by thorstenMueller** [reply]
Yes, increasing the value should help when output is cut off at the end.

**Reply by WeberJulian** [reply]
Except if you have very long test sentence it won't help you, it's probably your model alignment that is broken. Are test sentences sounding ok?

**Reply by thorstenMueller** [reply]
Indeed, @WeberJulian is right. How long are the phrases you try to synthesize?

**Reply by omkarade** [reply]
thanks for reply @thorstenMueller @WeberJulian

**Reply by sudip550** [reply]
> Indeed, @WeberJulian is right. How long are the phrases you try to synthesize?

sir please tell me, how to set up formatters.py. please help sir

**Comment by lalit-mohanani** [comment]
Hii @omkarade I am also facing the same issue. can you help me with that I am trying to train glowtts and tacotron2 with hindi dataset with ljspeech formator how can I modify recipies to train it for hindi.

**Comment by tan1abhi** [comment]
Hi , how can I use your model as a pre-trained model.
